src/cmap/connection.ts

```diff
@@ -1181,121 +1186,83 @@ export class ModernConnection extends TypedEventEmitter<ConnectionEvents> {
-    }
-
-    bufferPool.append(chunk);
-    const sizeOfMessage = bufferPool.getInt32();
+  /**
+   * @internal
+   *
+   * Writes an OP_MSG or OP_QUERY request to the socket, optionally compressing the command. This method
+   * waits until the socket's buffer has emptied (the Nodejs socket `drain` event has fired).
+   */
+  async writeCommand(
+    command: WriteProtocolMessageType,
+    options: Partial<Pick<OperationDescription, 'agreedCompressor' | 'zlibCompressionLevel'>>
+  ): Promise<void> {
+    const finalCommand =
+      options.agreedCompressor === 'none' || !OpCompressedRequest.canCompress(command)
+        ? command
+        : new OpCompressedRequest(command, {
+            agreedCompressor: options.agreedCompressor ?? 'none',
+            zlibCompressionLevel: options.zlibCompressionLevel ?? 0
+          });
 
-    if (sizeOfMessage == null) {
-      continue;
-    }
+    const buffer = Buffer.concat(await finalCommand.toBin());
 
-    if (sizeOfMessage < 0) {
-      throw new MongoParseError(`Invalid message size: ${sizeOfMessage}`);
-    }
+    return this.socketWrite(buffer);
+  }
 
-    if (sizeOfMessage > maxBsonMessageSize) {
-      throw new MongoParseError(
-        `Invalid message size: ${sizeOfMessage}, max allowed: ${maxBsonMessageSize}`
-      );
-    }
+  /**
+   * @internal
+   *
+   * Returns an async generator that yields full wire protocol messages from the underlying socket.  This function
+   * yields messages until `moreToCome` is false or not present in a response, or the caller cancels the request
+   * by calling `return` on the generator.
+   *
+   * Note that `for-await` loops call `return` automatically when the loop is exited.
+   */
+  async *readMany(): AsyncGenerator<OpMsgResponse | OpQueryResponse> {
+    for await (const message of onData(this.messageStream, { signal: this.controller.signal })) {
+      const response = await decompressResponse(message);
+      yield response;
 
-    if (sizeOfMessage > bufferPool.length) {
-      continue;
+      if (!response.moreToCome) {
+        return;
+      }
     }
-
-    yield bufferPool.read(sizeOfMessage);
   }
 }
 
-/**
- * @internal
- *
- * Writes an OP_MSG or OP_QUERY request to the socket, optionally compressing the command. This method
- * waits until the socket's buffer has emptied (the Nodejs socket `drain` event has fired).
- */
-export async function writeCommand(
-  connection: ModernConnection,
-  command: WriteProtocolMessageType,
-  options: Partial<Pick<OperationDescription, 'agreedCompressor' | 'zlibCompressionLevel'>> & {
-    signal?: AbortSignal;
-  }
-): Promise<void> {
-  const finalCommand =
-    options.agreedCompressor === 'none' || !OpCompressedRequest.canCompress(command)
-      ? command
-      : new OpCompressedRequest(command, {
-          agreedCompressor: options.agreedCompressor ?? 'none',
-          zlibCompressionLevel: options.zlibCompressionLevel ?? 0
-        });
+/** @internal */
```

Comment:
> going back to using a stream to just collect chunks gets us within a margin of error of the old performance

It was necessary, the original benchmarks show a mix of regressions but the worst case was a 30% regression, fixing the issue by making fewer abort signals gains about 10% back and adding a minimal stream regained the remaining 20%. 

I did attempt just using a connection.bufferPool with the data events from the socket but it still showed measurably worse performance, I suspect that creating a promise per chunk rather than per message is what keeps us on par with the old connection class. 

> I'd prefer not to use a custom stream if we can help it.

I am curious why we would avoid streams, they are generally the well-trodden path for working with Node.js I/O APIs.